File: Data Structures & Algorithms/subsets/submission-1.rs
impl Solution {
    pub fn subsets(nums: Vec<i32>) -> Vec<Vec<i32>> {
        fn dfs(
            nums: &[i32],
            index: usize,
            path: &mut Vec<i32>,
            result: &mut Vec<Vec<i32>>,
        ) {
            if index == nums.len() {
                result.push(path.clone());
                return;
            }

            dfs(nums, index + 1, path, result);
            path.push(nums[index]);
            dfs(nums, index + 1, path, result);
            path.pop();
        }

        let mut result = Vec::with_capacity(1usize << nums.len());
        let mut path = Vec::with_capacity(nums.len());
        dfs(&nums, 0, &mut path, &mut result);
        result
    }
}

File: Data Structures & Algorithms/permutations/submission-1.py
from typing import List


class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        values = nums.copy()
        result = []

        def dfs(start):
            if start == len(values):
                result.append(values.copy())
                return
            for i in range(start, len(values)):
                values[start], values[i] = values[i], values[start]
                dfs(start + 1)
                values[start], values[i] = values[i], values[start]

        dfs(0)
        return result